src/TypeScript/TsConfig/TsConfigProvider.cs

```diff
@@ -61,16 +62,69 @@ public async Task<string> GetConfigForFile(string sourceFilePath, CancellationTo
 
             logger.LogDebug(Resources.INFO_FoundTsConfigs, string.Join(Path.PathSeparator.ToString(), allTsConfigsFilePaths));
 
-            foreach (var tsConfigsFilePath in allTsConfigsFilePaths)
+            var tsConfigFile = await GetConfigForFile(sourceFilePath,
+                allTsConfigsFilePaths,
+                checkedTsConfigs: new List<string>(),
+                cancellationToken);
+
+            return tsConfigFile;
+        }
+
+        private async Task<string> GetConfigForFile(string sourceFilePath,
+            IEnumerable<string> candidateTsConfigs,
+            ICollection<string> checkedTsConfigs,
+            CancellationToken cancellationToken)
+        {
+            foreach (var tsConfigFilePath in candidateTsConfigs)
             {
-                var response = await typeScriptEslintBridgeClient.TsConfigFiles(tsConfigsFilePath, cancellationToken);
+                if (checkedTsConfigs.Contains(tsConfigFilePath))
+                {
+                    continue;
+                }
+
+                checkedTsConfigs.Add(tsConfigFilePath);
+
+                var response = await typeScriptEslintBridgeClient.TsConfigFiles(tsConfigFilePath, cancellationToken);
+
+                if (response.Error != null)
+                {
+                    logger.WriteLine(Resources.ERR_FailedToProcessTsConfig, tsConfigFilePath, response.Error);
+                    continue;
+                }
+
+                if (response.ParsingError != null)
+                {
+                    logger.WriteLine(Resources.ERR_FailedToProcessTsConfig_ParsingError, 
+                        tsConfigFilePath, 
+                        response.ParsingError.Code,
+                        response.ParsingError.Line,
+                        response.ParsingError.Message);
+                    continue;
+                }
 
                 if (response.Files != null &&
-                    response.Files.Any(x => IsMatchingPath(x, sourceFilePath, tsConfigsFilePath)))
+                    response.Files.Any(x => IsMatchingPath(x, sourceFilePath, tsConfigFilePath)))
                 {
-                    logger.WriteLine(Resources.INFO_MatchingTsConfig, sourceFilePath, tsConfigsFilePath);
+                    logger.WriteLine(Resources.INFO_MatchingTsConfig, sourceFilePath, tsConfigFilePath);
 
-                    return tsConfigsFilePath;
+                    return tsConfigFilePath;
+                }
+
+                if (response.ProjectReferences == null || !response.ProjectReferences.Any())
+                {
+                    continue;
+                }
+
+                logger.LogDebug(Resources.INFO_CheckingReferencedTsConfigs, tsConfigFilePath, string.Join(Path.DirectorySeparatorChar.ToString(), response.ProjectReferences));
+
+                var result = await GetConfigForFile(sourceFilePath,
+                    response.ProjectReferences,
```

Comment:
On the java side, they first gather all the tsconfigs and their references, and only then iterate over them to find one that contains the analyzed file.
Implementing it this way would help with #2446, but it would also hurt the performance. In this PR everything is evaluated lazily. I'm a bit reluctant to take the java approach as I've seen eslint-bridge take 4+ seconds to return a result for a single tsconfig.
For #2446, perhaps we can start by sorting and giving priority to the closest tsconfig found in the project's directory, ignoring the project-references. Perhaps the tsconfig found in the project directory be closer than the tsconfigs mentioned in projectReferences, but I don't know what are the real-world scenarios. WDYT?

Reply:
The Java code is in the analyzer sensor, which is optimised for the batch processing scanning performed by the scanner. The other SL flavours don't have any opportunity to change the approach, since they are emulating the scanner API.
It doesn't make sense for us to use the same approach for single file analysis.

I don't know enough about the real-world scenarios either.

Scenario:
* tsconfig **A** is closest to the file
* **A** contains settings but doesn't contain any includes
* **A** referencs tsconfig **B**
* **B** includes the file.

Which tsconfig should we return? **A**, since it's closer and indirectly references the file, or **B** which directly references the file?

Reply:
I honestly don't know. I'd say that this is a very weird project structure 😆